modules/scrollback/index.js

```diff
@@ -110,108 +111,389 @@ module.exports.formatLine = function(line) {
-          lineObj.regex = regex;
-        } catch(ex) {
-          return cb("Could not parse regex: " + parts[i] + ", " + ex.toString());
+      else {
+        if(i != result.index) { // Got some lines
+          i = result.index;
+          spec.lines = spec.lines.concat(result.result);
         }
-      } else {
-        lineObj.lines = [1];
-        i--;
-        // Backtrack, nick not followed by anything it turns out
       }
-      linesToGet.push(lineObj);
     }
+    else if(state == 'new') {
+      specs.push(spec);
+      spec = {nicks: [], regexes: [], lines: []};
+    }
+
+    else if(state == 'error') {
+      return {error: error};
+    }
+
+    // Get next state
+    result = intuitState(input, i, state);
+    i = result.index;
+    state = result.state;
+    if(result.error !== undefined) {
+      error = result.error;
+    }
+  }
+
+  specs.push(spec);
+
+  return _.filter(specs, function(spec) {
+    return (spec.nicks.length > 0) ||
+      (spec.regexes.length > 0) ||
+      (spec.lines.length > 0);
+  });
+}
+
+function parseNick(str, idx) {
+  var ret = "";
+
+  while(idx < str.length && str[idx] != ' ' && str[idx] != ',') {
+    ret += str[idx];
+    idx++;
+  }
+
+  return {result: ret, index: idx};
+}
+
+function parseRegex(str, idx) {
+  var ret = "";
+  var regex;
+
+  idx++; // Eat leading /
+
+  while(idx < str.length && str[idx] != '/') {
+    ret += str[idx];
+    if(idx < str.length-1 && str[idx] === '\\') { // Skip over escapes
+      idx++;
+      ret += str[idx];
+    }
+
+    idx++;
+  }
+
+  if(idx < str.length) { // Catch the trailing /
+    idx++;
+
+    try { // Parse the regex
+      regex = new RegExp(ret);
+      return {result: regex, index: idx};
+    } catch(ex) {
+      return {result: ret, index: idx, error: "Could not parse regex: " + ret + "/, " + ex.toString()};
+    }
+  }
+  else { // Unterminated regex
+    return {result: ret, index: idx, error: "Unterminated regular expression"};
+  }
+}
+
+function parseLines(str, idx) {
+  var ret = "";
+
+  while(idx < str.length && /[0-9\-.,]/.test(str[idx])) {
+    ret += str[idx];
+    idx++;
+  }
+
+  if(str[idx-1] === ',') { // Put back trailing commas
+    idx--;
+    ret = ret.slice(0,-1);
+  }
+
+  var range = rangeParser.parse(ret);
+  if(range === undefined || range.length == 0) {
+    return {result: range, index: idx, error: "Could not parse " + ret};
+  }
+  if(_.every(range, function(num) { return num > 0; })) {
+    return {result: range, index: idx};
+  }
+  else {
+    return {result: range, index: idx, error: "Negative line index"};
+  }
+}
+
+function intuitState(str, idx, state) {
+  var isList = false;
+  var next = {};
+
+  if(state == 'error') {
+    return {state: state, index: idx};
+  }
+
+  while(idx < str.length && str[idx] == ' ') { // Consume whitespace
+    idx++;
+  }
+
+  if(idx < str.length-1 && str[idx] == ',') {
+    if(str[idx+1] == ' ') { // ", " condition
+      idx++;
+      return {state:'new', index: idx};
```

Comment:
This is a nit, but the code would probably read better with:

* An enum-object (`ParserStates.NEW = 'new'` stuff)
* Possibly an object to represent this stuff beyond a dumb dict.

I'm fine with neither of those things being done, but I do think it'd make it a little more readable